### PR TITLE
Add benchmark for batch gen/eval

### PR DESCRIPTION
**Pull request by arturomf94**
## Description
We've already got some benchmarks but we're missing one for batch gen/eval. This PR adds that.

## Affected Dependencies
.

## How has this been tested?
Running benchmarks locally.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: benches/bench_eq.rs [hunk 1]
```diff
@@ -27,5 +44,39 @@ pub fn eq_eval(c: &mut Criterion) {
     });
 }
 
-criterion_group!(bench_eq, eq_keygen, eq_eval);
+pub fn eq_batch_eval(c: &mut Criterion) {
+    let mut keys_a: Vec<u8> = Vec::with_capacity(3105);
```
Comment (tholop) [inline benches/bench_eq.rs:48]: Please add a comment to recap where this magic number comes from (either with the math or with a link to some code that already contains this number)

File: benches/bench_eq.rs [hunk 2]
```diff
@@ -27,5 +44,39 @@ pub fn eq_eval(c: &mut Criterion) {
...
+    let mut keys_b: Vec<u8> = Vec::with_capacity(3105);
+    let keys_a_pointer: *mut u8 = keys_a.as_mut_ptr();
+    let keys_b_pointer: *mut u8 = keys_b.as_mut_ptr();
+    let n_values: usize = 5;
```
Comment (tholop) [inline benches/bench_eq.rs:52]: As you pointed out by MP, I think that we need multiple benches with different parameters :) This one is fine, but we should evaluate batch keygen for larger batches: the whole point of this batch generation is to be efficient when we preprocess large number of keys. n_values = 100 or 1000 or even 10_000 would be great, if it is fast enough.

File: benches/bench_eq.rs [hunk 3]
```diff
@@ -27,5 +44,39 @@ pub fn eq_eval(c: &mut Criterion) {
...
+    let n_threads: usize = 6;
```
Comment (tholop) [inline benches/bench_eq.rs:53]: I don't know how many threads are going to be used by the Github Actions worker. Anyway, it is probably worth launching multiple benches with n_threads = 1, 4, 6 for instance

File: benches/bench_eq.rs [hunk 4]
```diff
@@ -14,6 +15,22 @@ pub fn eq_keygen(c: &mut Criterion) {
...
+pub fn eq_batch_keygen(c: &mut Criterion) {
+    let mut keys_a: Vec<u8> = Vec::with_capacity(3105);
+    let mut keys_b: Vec<u8> = Vec::with_capacity(3105);
+    let keys_a_pointer: *mut u8 = keys_a.as_mut_ptr();
+    let keys_b_pointer: *mut u8 = keys_b.as_mut_ptr();
+    let n_values: usize = 5;
+    let n_threads: usize = 6;
+    let op_id: usize = 0;
+    unsafe {
+        c.bench_function("Batch keygen", |b| {
```
Comment (tholop) [inline benches/bench_eq.rs:28]: Nit: "Batch Eq keygen"

File: benches/bench_le.rs [hunk 5]
```diff
@@ -14,6 +15,22 @@ pub fn le_keygen(c: &mut Criterion) {
...
+pub fn le_batch_keygen(c: &mut Criterion) {
+    let mut keys_a: Vec<u8> = Vec::with_capacity(4600);
+    let mut keys_b: Vec<u8> = Vec::with_capacity(4600);
+    let keys_a_pointer: *mut u8 = keys_a.as_mut_ptr();
+    let keys_b_pointer: *mut u8 = keys_b.as_mut_ptr();
+    let n_values: usize = 5;
+    let n_threads: usize = 6;
+    let op_id: usize = 1;
+    unsafe {
+        c.bench_function("Batch keygen", |b| {
```
Comment (tholop) [inline benches/bench_le.rs:28]: Nit: add "Le"

**Review by tholop — CHANGES_REQUESTED** [review]
Thanks a lot, @arturomf94! The overall structure looks great. I left a couple of comments, mostly to add more benches with different configurations.

Regarding your comment about the duration: indeed, a batch keygen/eval takes longer than a single keygen/eval. But when we normalize by the total number of keys (`n_values`), we obtain a lower per-key time with the batch functions. I expect that gap to be even wider when we use higher values of `n_values` instead of 5 or 6 :)

File: benches/bench_eq.rs (hunk 3 above)
Comment (arturomf94) [reply benches/bench_eq.rs:57]: [Here](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners) is some info on the specifications of the GitHub Actions runners. I don't know how many threads per core they have but we could certainly add the benchmarks.

**Comment by arturomf94** [comment]
@tholop - attending your comments, I've grouped the benchmarks with ranges of `n_values` and `n_threads`. I'm not sure how this will play out in the GitHub Actions runners (see info [here](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners)) but hopefully this is closer to what will be useful for us. Let me know if there's anything missing.

**Comment by tholop** [comment]
Looks perfect, thanks @arturomf94! I realized that the speed was the same regardless of the value of `n_threads` -- not because of the benchmark code. The reason is that if a Rayon thread pool already exists, Sycret will not initialize a new one: https://github.com/OpenMined/sycret/blob/fc776e1a9c517aebdd7a705d448eb1865cda6479/src/lib.rs#L84

Most of the time, the user will not provide any value for `n_threads anyway`, so Sycret will directly use the right number of threads, but this is not very clean. We don't want to pay the Rayon pool initialization overhead at every call, but I think that we can respawn the pool only when `n_threads` is different than `rayon::current_num_threads`. What do you think? 

I'm happy to push a commit for this change.

**Comment by arturomf94** [comment]
> We don't want to pay the Rayon pool initialization overhead at every call, but I think that we can respawn the pool only when n_threads is different than rayon::current_num_threads

Yes - I think this sounds optimal.

**Pull request merged**